pkg/spdk/backing_image.go

```diff
@@ -0,0 +1,735 @@
+	bi.UpdateCh <- nil
+}
+
+func (bi *BackingImage) updateProgress(processedSize int64) {
+	bi.Lock()
+	defer bi.Unlock()
+
+	if bi.State == types.BackingImageStateStarting {
+		bi.State = types.BackingImageStateInProgress
+	}
+	if bi.State == types.BackingImageStateReady {
+		return
+	}
+
+	bi.ProcessedSize = bi.ProcessedSize + processedSize
+	if bi.Size > 0 {
+		bi.Progress = int32((float32(bi.ProcessedSize) / float32(bi.Size)) * 100)
+	}
+}
+
+func (bi *BackingImage) prepareBackingImageSnapshot(spdkClient *spdkclient.Client, superiorPortAllocator *commonbitmap.Bitmap, fromAddress string, srcLvsUUID string) (err error) {
+	// If we already create the temp head lvol, no matter it fails to prepare or not, we should create the snapshot for it.
+	// The state will be recorded in the xattr of the snapshot lvol so when the node is rebooted, we can pick it up and reconstruct the record.
+	biTempHeadUUID := ""
+
+	defer func() {
+		bi.Lock()
+		defer bi.Unlock()
+
+		if bi.IsExposed {
+			bi.log.Info("Unexposing lvol bdev")
+			backingImageTempHeadName := GetBackingImageTempHeadLvolName(bi.Name, bi.LvsUUID)
+			if err = spdkClient.StopExposeBdev(helpertypes.GetNQN(backingImageTempHeadName)); err != nil {
+				logrus.Errorf("Failed to unexpose lvol bdev %v after failed to create backing image", backingImageTempHeadName)
+			}
+			bi.IsExposed = false
+		}
+		if bi.Port != 0 {
+			if err := superiorPortAllocator.ReleaseRange(bi.Port, bi.Port); err != nil {
+				logrus.Errorf("Failed to release port %v after failed to create backing image", bi.Port)
+			}
+			bi.Port = 0
+		}
+
+		if err == nil {
+			if bi.State != types.BackingImageStateInProgress {
+				err = fmt.Errorf("invalid state %v for backing image %s creation after processing", bi.State, bi.Name)
+			}
+
+			if bi.Size > 0 && bi.ProcessedSize != int64(bi.Size) {
+				err = fmt.Errorf("processed data size %v does not match the expected file size %v", bi.ProcessedSize, bi.Size)
+			}
+
+			if bi.CurrentChecksum != bi.ExpectedChecksum {
+				err = fmt.Errorf("current checksum %v does not match the expected checksum %v", bi.CurrentChecksum, bi.ExpectedChecksum)
+			}
+		}
+
+		if err != nil {
+			bi.log.WithError(err).Errorf("Failed to create backing image %v", bi.Name)
+			if bi.State != types.BackingImageStateFailed {
+				bi.State = types.BackingImageStateFailed
+			}
+			bi.ErrorMsg = err.Error()
+		} else {
+			bi.Progress = 100
+			bi.State = types.BackingImageStateReady
+			bi.ErrorMsg = ""
+		}
+
+		if biTempHeadUUID != "" {
+			// Create a snapshot with backing image name from the backing image temp head lvol
+			// the snapshot lvol name will be "bi-${biName}-disk-${lvsUUID}"
+			// and the alias will be "${lvsName}/bi-${biName}-disk-${lvsUUID}"
+
+			ne, err := helperutil.NewExecutor(commontypes.HostProcDirectory)
+			if err != nil {
+				bi.log.WithError(err).Errorf("WARNING: failed to get the executor for snapshot backing image %v, will skip the sync and continue", bi.Name)
```

Comment:
Using warning message is good enough
```suggestion
				bi.log.WithError(err).Warnf("Failed to get the executor for snapshot backing image %v, will skip the sync and continue", bi.Name)
```

Reply:
fixed.